[Incomplete] Feature: Snackbar implementation

gugu opened the pull request:
The initial version of Snackbar. If code will be in general OK, I'll add tests as well

If you want to try: `npm install @shortcm/snackbar`

moog16 commented on packages/snackbar/index.tsx:
```diff
@@ -0,0 +1,143 @@
+// The MIT License
+//
+// Copyright (c) 2018 Google, Inc.
```
2019

moog16 commented on packages/snackbar/package.json:
```diff
@@ -0,0 +1,28 @@
+{
+  "name": "@shortcm/react-snackbar",
```
can we change this to material?

moog16 commented on packages/snackbar/index.tsx:
```diff
@@ -0,0 +1,143 @@
+        });
+      },
+      announce: () => {
+        console.log('TODO Snackbar/announce - have no idea what to put here');
```
this could probably be another callback method set through props

moog16 commented on packages/snackbar/index.tsx:
```diff
@@ -0,0 +1,167 @@
+// @ts-ignore no .d.ts file
+import {MDCSnackbarFoundation} from '@material/snackbar';
+
+type Props = {
```
export these Props...Can you also convert to interface?

moog16 commented on packages/snackbar/index.tsx:
```diff
@@ -0,0 +1,167 @@
+        });
+      },
+      announce: () => {
+        // Usually it works automatically if this component uses conditional rendering
```
What does this comment mean?

gugu replied:
In MDC Snackbar component there are some hacks to make `announce` work correctly.
With React, it works correctly out of the box. As soon as it renders initially, screenreader announces the text

moog16 commented on packages/snackbar/README.md:
```diff
@@ -0,0 +1,64 @@
+
+## Props
+
+Prop Name | Type | Description
```
add onAnnounce to readme

moog16 commented on packages/snackbar/index.tsx:
```diff
@@ -0,0 +1,167 @@
+    return classnames(this.props.className, ...Array.from(this.state.classes));
+  }
+  render() {
+    return <div className={`mdc-snackbar ${this.classes}`} onKeyDown={this.handleKeyDown}>
```
Can you add `mdc-snackbar` to `this.classes()`? Then this line is just `className={this.classes}`

JoelSweeneySlalom replied:
But you could have 'mdc-snackbar' as the first argument passed to classnames in the classes getter, then it would always be first, would it not?

gugu replied:
You are right

moog16 commented on packages/snackbar/package.json:
```diff
@@ -0,0 +1,28 @@
+{
+  "name": "@material/react-snackbar",
+  "version": "0.8.3",
```
change this to 0.0.0

moog16 commented on packages/snackbar/package.json:
```diff
@@ -0,0 +1,28 @@
+    "url": "https://github.com/material-components/material-components-web-react.git"
+  },
+  "dependencies": {
+    "@material/button": "^0.41.0",
```
remove @material/button and @material/react-ripple. This isn't used in the component

moog16 commented on packages/snackbar/types.tsx:
```diff
@@ -0,0 +1,24 @@
+export interface IMDCSnackbarFoundation {
```
add MIT license here.

Also please add a comment saying:
```
// TODO: remove this when MDC Web types are added.
```

moog16 commented on scripts/package-json-reader.js:
```diff
@@ -10,12 +10,12 @@ const readMaterialPackages = () => {
 
   for (let dep in packageJson.devDependencies) {
-    if (dep.startsWith('@material/')) {
+    if (dep.startsWith('@material/') || dep.startsWith('@shortcm/')) {
```
revert all changes in this file.

moog16 commented on test/screenshot/snackbar/index.scss:
```diff
@@ -0,0 +1,6 @@
+.snackbar-container {
+  margin-bottom: 16px;
+  .mdc-snackbar {
+    position: static !important;
```
I know this is just a screenshot tests, but the `!important` is a code smell. Why is this needed?

gugu replied:
By default, snack bar uses position: fixed and second item is displayed over the first one. I have less bad solution for this

moog16 replied:
ah...sounds like an mdc web issue then?

gugu replied:
no, it is correct behavior for the browser. But for screenshot tests it will not allow to test snackbar correctly. `.mdc-snackbar.mdc-snackbar` can be used instead of !important

moog16 commented on test/unit/snackbar/index.test.tsx:
```diff
@@ -0,0 +1,35 @@
+import * as React from 'react';
```
Need tests for adapter methods.

gugu replied:
done

moog16 commented:
fixes https://github.com/material-components/material-components-web-react/issues/447

moog16 commented on test/screenshot/snackbar/index.scss:
```diff
@@ -1,6 +1,6 @@
   margin-bottom: 16px;
-  .mdc-snackbar {
-    position: static !important;
+  .mdc-snackbar.mdc-snackbar {
```
ya - not great, but i like it better than !important :)

moog16 replied:
can you just add a comment as to why you need this.

moog16 replied:
thanks!

moog16 commented on packages/snackbar/index.tsx:
```diff
@@ -24,20 +24,21 @@ import * as React from 'react';
   leading?: boolean,
   stacked?: boolean,
-  onBeforeOpen?: () => void,
+  onOpening?: () => void,
```
change all the `,` --> `;`

moog16 commented:
@gugu I see you did more tests! If you run `npm run test:unit`, this will create a `coverage` directory. Can you make sure that branches, statements, lines, and functions is +95%. If you can't do it let me know. From the looks of it you're around 77%.

gugu commented:
Done. It now says about CLA signing, probably after merge

gugu commented:
The problem was when the base branch was changed. I removed extra commits and squashed my changes to one commit (to not to pollute module history). Also I added karma config instead of tests. Now test coverage is 100%

moog16 commented on packages/snackbar/README.md:
````diff
@@ -0,0 +1,65 @@
+  }
+}
+```
+
````
Can you also add documentation on how to use `getTimeoutMs`, `getCloseOnEscape`, and `isOpen`? Something like:

```
If you need to get the `timeoutMs`, `closeOnEscape`, or `open` value, then you can use a ref like so:

import React from 'react';
import Snackbar from '@material/react-snackbar';
 class MyApp extends React.Component {
  getSnackbarInfo = (snackbar) => {
    if (!snackbar) return;
    console.log(snackbar.getTimeoutMs());
    console.log(snackbar.isOpen());
    console.log(snackbar.getCloseOnEscape());
  }
  render() {
    return (
      <Snackbar
        message="Click Me!"
        actionText="dismiss"
        ref={this.getSnackbarInfo}
      />
    );
  }
}
```

moog16 commented:
I realized that you also need to add snackbar to the `screenshot-test-urls.tsx` file, and water with rc0.10.0.

gugu commented:
Done with docs and screenshot-test-urls. What does "water with rc0.10.0." mean? I did rebase from rc0.10.0, but I'm not sure it is what did you mean

moog16 commented:
water === merge...the more and more I say that, the more I think it is not a real term. Rebase is fine :)

moog16 commented on packages/snackbar/README.md:
```diff
@@ -0,0 +1,89 @@
+
+If you need to get the `timeoutMs`, `closeOnEscape`, or `open` value, then you can use a ref like so:
+
+import React from 'react';
```
could you put this in a code block using ` ``` `

moog16 commented:
@gugu created a screenshot. please update snackbar golden.json file:
`fb7f6f61f37095d7e7c92c9305eb797145d8e72c0c8dbc4e531f9aca91b0fd28`

moog16 commented:
i made the gc bucket public (I think)...try running `npm run lint`.

moog16 commented:
tests for https://github.com/material-components/material-components-web-react/pull/637

moog16 commented:
@gugu  did you sign PR...i remember you doing so.

moog16 commented:
Thanks :)

gugu commented:
rebased to the last rc0.10.0 commit
